File: src/main/java/axirassa/webapp/services/EntityRealm.java

package axirassa.webapp.services;

import org.apache.shiro.authc.AccountException;
import org.apache.shiro.authc.AuthenticationException;
import org.apache.shiro.authc.AuthenticationInfo;
import org.apache.shiro.authc.AuthenticationToken;
import org.apache.shiro.authc.UsernamePasswordToken;
import org.apache.shiro.authz.AuthorizationInfo;
import org.apache.shiro.authz.SimpleAuthorizationInfo;
import org.apache.shiro.cache.MemoryConstrainedCacheManager;
import org.apache.shiro.realm.AuthorizingRealm;
import org.apache.shiro.subject.PrincipalCollection;

import axirassa.dao.UserEmailAddressDAO;
import axirassa.model.UserEntity;

/**
 * Based on suggestion from
 * http://permalink.gmane.org/gmane.comp.java.tynamo.user/155
 * 
 * @author wiktor
 */
public class EntityRealm extends AuthorizingRealm {
	public static final String REALM_NAME = "axirassarealm";

	private final UserEmailAddressDAO emailDAO;


	public EntityRealm(UserEmailAddressDAO emailDAO) {
		super(new MemoryConstrainedCacheManager());

		this.emailDAO = emailDAO;
		setName(REALM_NAME);
		setAuthenticationTokenClass(UsernamePasswordToken.class);
	}


	@Override
	protected AuthorizationInfo doGetAuthorizationInfo(PrincipalCollection principals) {
		if (principals.isEmpty())
			return null;

		// make sure we have some authorization info from this realm
		if (principals.fromRealm(REALM_NAME).isEmpty())
			return null;

		String email = (String) principals.fromRealm(REALM_NAME).iterator().next();

		// no e-mail, no credentials
		if (email == null)
			return null;

		UserEntity user = emailDAO.getUserByEmail(email);
		if (user == null)
			return null;

		return new SimpleAuthorizationInfo(user.roles());
	}


	@Override
	protected AuthenticationInfo doGetAuthenticationInfo(AuthenticationToken token) throws AuthenticationException {
		UsernamePasswordToken uptoken = (UsernamePasswordToken) token;

		String email = uptoken.getUsername();

		if (email == null)
			throw new AccountException("empty username for realm: " + REALM_NAME);

		// verify account exists
		UserEntity user = emailDAO.getUserByEmail(email);

		// retrieve the password and salt
		byte[] password = user.getPassword();
		byte[] salt = user.getSalt();

		return new UserAuthenticationInfo(email, password, salt);
	}

}

File: src/main/java/axirassa/webapp/services/EntityRealm.scala
package axirassa.webapp.services

import org.apache.shiro.authc.AccountException
import org.apache.shiro.authc.UsernamePasswordToken
import org.apache.shiro.authc.AuthenticationToken

import axirassa.dao.UserDAO
import axirassa.model.UserEntity 
import org.apache.shiro.authz.AuthorizationInfo
import org.apache.shiro.authz.SimpleAuthorizationInfo
import org.apache.shiro.cache.MemoryConstrainedCacheManager
import org.apache.shiro.realm.AuthorizingRealm
import org.apache.shiro.subject.PrincipalCollection


object EntityRealm {
    val REALM_NAME = "axirassarealm"
}

/**
 * Based on suggestion from http://permalink.gmane.org/gmane.comp.java.tynamo.user/155
 *
 * @author wiktor
 */
class EntityRealm(userDAO : UserDAO) extends AuthorizingRealm(new MemoryConstrainedCacheManager()) {	    
	setName(EntityRealm.REALM_NAME)
	setAuthenticationTokenClass(classOf[UsernamePasswordToken])


	override protected
	def doGetAuthorizationInfo (principals : PrincipalCollection ) {
		if (principals.isEmpty())
			return null

		// make sure we have some authorization info from this realm
		if (principals.fromRealm(EntityRealm.REALM_NAME).size() <= 0)
			return null

		val email = principals.fromRealm(EntityRealm.REALM_NAME).iterator().next().asInstanceOf[String]

		// no e-mail, no credentials
		if (email == null)
			return null

		val user = userDAO.getUserByEmail(email)
		if (user.isEmpty)
			return null

		return new SimpleAuthorizationInfo(user.get.roles())
	}


	protected override
	def doGetAuthenticationInfo ( token : AuthenticationToken) = {
		val uptoken = token.asInstanceOf[UsernamePasswordToken] 

		val email = uptoken.getUsername()

		if (email == null)
			throw new AccountException("empty username for realm: " + EntityRealm.REALM_NAME)

		// verify account exists
		val user = userDAO.getUserByEmail(email)

		// retrieve the password and salt
		val password = user.get.getPassword()
		val salt = user.get.getSalt()

	  new UserAuthenticationInfo(email, password, salt)
	}

}
